src/navigation/failure_identification/watchdog.py

```diff
@@ -1,11 +1,130 @@
 import numpy as np
 import rospy
 from pandas import DataFrame
+from util.ros_utils import get_rosparam
+from util.SO3 import SO3
+
+DF_THRESHOLD = get_rosparam("watchdog/dataframe_threshold", 100)
+STUCK_THRESHOLD = get_rosparam("watchdog/stuck_threshold", 50)
+ANGULAR_THRESHOLD = get_rosparam("watchdog/angular_threshold", 0.001)
+LINEAR_THRESHOLD = get_rosparam("watchdog/linear_threshold", 0.55)
 
 
 class WatchDog:
-    def __init__(self):
-        pass
+    def __init__(self, collector_in):
+        self.collector = collector_in
+
+    def update_pointers(self):
+        self.collector.right_pointer = self.collector.row_counter
+        self.collector.left_pointer = self.collector.right_pointer - DF_THRESHOLD
+
+    def get_start_end_positions(self, dataframe: DataFrame):
+        #get the average of the first 25% of the dataframes x, y, z position from the inputted dataframe
+        cutoff = int(len(dataframe) * 0.25)
+        start = dataframe.loc[dataframe["row"] <= cutoff]
+        end = dataframe.loc[dataframe["row"] >= len(dataframe) - cutoff]
+        start_x = np.mean(start["x"])
+        start_y = np.mean(start["y"])
+        start_z = np.mean(start["z"])
+        start_pos = np.array([start_x, start_y, start_z])
+        end_x = np.mean(end["x"])
+        end_y = np.mean(end["y"])
+        end_z = np.mean(end["z"])
+        end_pos = np.array([end_x, end_y, end_z])
+        return start_pos, end_pos
+    
+    def get_start_end_rotations(self, dataframe: DataFrame):
+        #get the average of the first 25% of the dataframes rotation from the inputted dataframe
+        cutoff = int(len(dataframe) * 0.25)
+        start = dataframe.loc[dataframe["row"] <= cutoff]
+        end = dataframe.loc[dataframe["row"] >= len(dataframe) - cutoff]
+        start_rot = np.array([np.mean(start["rot_x"]), np.mean(start["rot_y"]), np.mean(start["rot_z"]), np.mean(start["rot_w"])])
+        end_rot = np.array([np.mean(end["rot_x"]), np.mean(end["rot_y"]), np.mean(end["rot_z"]), np.mean(end["rot_w"])])
+        return SO3(start_rot), SO3(end_rot)
+        
+    def get_start_end_time(self, dataframe: DataFrame):
+        #get the average of the first 25% of the dataframes time from the inputted dataframe
+        cutoff = int(len(dataframe) * 0.25)
+        start = dataframe.loc[dataframe["row"] <= cutoff]
```

Comment:
an example for iloc would be to change this to dataframe.iloc[:cutoff] (I realize this is my Github Copilot work lol)